relentless/optimize/criteria.py

```diff
@@ -0,0 +1,128 @@
+import abc
+
+import numpy as np
+
+class ConvergenceTest(abc.ABC):
+    """
+        abs_tol : float or dict
+            The absolute tolerance or tolerances (keyed on the :class:`~relentless.optimize.objective.ObjectiveFunction`
+            design variables).
+        rel_tol : float or dict
+            The relative tolerance or tolerances (keyed on the :class:`~relentless.optimize.objective.ObjectiveFunction`
+            design variables). (Defaults to ``None``).
+
+    """
+    def __init__(self, abs_tol, rel_tol=None, mode='or'):
```

Comment:
Overall, I very much like this design, but I think this is a little too base-class heavy. I don't think all `ConvergenceTest`s should be required to support both absolute tolerance and relative tolerance or these different modes.

Rather, I would make `ConvergenceTest` be a light class:
```py
class ConvergenceTest(abc.ABC):
    @abc.abstractmethod
    def converged(self, result):
        pass
```

Then, you can subclass abstract classes for absolute and relative tolerances:
```py
class AbsoluteTolerance(ConvergenceTest):
    def __init__(self, tolerance):
        pass

class RelativeTolerance(ConvergenceTest):
    def __init__(self, tolerance):
        pass
```
and implement specific cases of each:
```py
class GradientAbsoluteTolerance(AbsoluteTolerance):
    pass
class GradientRelativeTolerance(RelativeTolerance):
    pass
```

In this way, `LogicTest` can also subclass from `ConvergenceTest`:
```py
class LogicTest(ConvergenceTest):
    pass
``
I think we voted not to do component-wise and/or tests, so you can remove mode an implement this with the `OrTest` and `AndTest`.